test/functional/zmq_test.py

```diff
@@ -38,72 +52,71 @@ def setup_nodes(self):
+        assert_equal(msgSequence, 0)  # must be sequence 0 on hashtx
 
-        msg = self.zmqSubSocket.recv_multipart()
+        self.log.info("Wait for block")
+        msg = wait_for_multipart(self.zmqSubSocket)
         topic = msg[0]
         body = msg[1]
         msgSequence = struct.unpack('<I', msg[-1])[-1]
-        assert_equal(msgSequence, 0) #must be sequence 0 on hashblock
+        assert_equal(msgSequence, 0)  # must be sequence 0 on hashblock
         blkhash = bytes_to_hex_str(body)
 
-        assert_equal(genhashes[0], blkhash) #blockhash from generate must be equal to the hash received over zmq
+        assert_equal(genhashes[0], blkhash)  # blockhash from generate must be equal to the hash received over zmq
 
+        self.log.info("Generate 10 blocks (and 10 coinbase txes)")
         n = 10
         genhashes = self.nodes[1].generate(n)
         self.sync_all()
 
         zmqHashes = []
         blockcount = 0
-        for x in range(0,n*2):
-            msg = self.zmqSubSocket.recv_multipart()
+        for x in range(n * 2):
+            msg = wait_for_multipart(self.zmqSubSocket)
             topic = msg[0]
             body = msg[1]
             if topic == b"hashblock":
                 zmqHashes.append(bytes_to_hex_str(body))
                 msgSequence = struct.unpack('<I', msg[-1])[-1]
-                assert_equal(msgSequence, blockcount+1)
+                assert_equal(msgSequence, blockcount + 1)
                 blockcount += 1
 
-        for x in range(0,n):
-            assert_equal(genhashes[x], zmqHashes[x]) #blockhash from generate must be equal to the hash received over zmq
+        for x in range(n):
+            assert_equal(genhashes[x], zmqHashes[x])  # blockhash from generate must be equal to the hash received over zmq
 
-        #test tx from a second node
+        self.log.info("Wait for tx from second node")
+        # test tx from a second node
         hashRPC = self.nodes[1].sendtoaddress(self.nodes[0].getnewaddress(), 1.0)
         self.sync_all()
 
         # now we should receive a zmq msg because the tx was broadcast
-        msg = self.zmqSubSocket.recv_multipart()
+        msg = wait_for_multipart(self.zmqSubSocket)
         topic = msg[0]
         body = msg[1]
-        hashZMQ = ""
-        if topic == b"hashtx":
-            hashZMQ = bytes_to_hex_str(body)
-            msgSequence = struct.unpack('<I', msg[-1])[-1]
-            assert_equal(msgSequence, blockcount+1)
+        assert_equal(topic, b"hashtx")
+        hashZMQ = bytes_to_hex_str(body)
+        msgSequence = struct.unpack('<I', msg[-1])[-1]
+        assert_equal(msgSequence, blockcount + 1)
 
-        assert_equal(hashRPC, hashZMQ) #blockhash from generate must be equal to the hash received over zmq
+        assert_equal(hashRPC, hashZMQ)  # txid from sendtoaddress must be equal to the hash received over zmq
 
+        # Destroy the zmq context
+        self.zmqContext.destroy(linger=None)
```

Comment:
Good point. Added a fixup commit 6cdfd7020915356bd82c1f538c0b7817665fda20. Can you take a look and let me know what you think. If you're happy I'll squash with the previous commit.